File: illumenate_lighting/illumenate_lighting/api/product_projection.py
"""Shared schema-accurate product projection; no database or actor side effects."""

import json
from urllib.parse import unquote, urlencode, urlsplit

from illumenate_lighting.illumenate_lighting.api.configuration_contract import FAMILY_ALIASES, parse_bool

TEMPLATE_FIELDS = {
	"Linear Fixture": "fixture_template",
	"LED Tape": "tape_neon_template",
	"LED Neon": "tape_neon_template",
	"LED Sheet": "led_sheet_template",
}


def safe_document_url(value):
	if not isinstance(value, str) or not value or any(ord(c) < 32 for c in value):
		return None
	try:
		parts = urlsplit(value)
	except ValueError:
		return None
	if (
		value.startswith("/files/")
		and not parts.netloc
		and ".." not in unquote(parts.path).replace("\\", "/").split("/")
	):
		return value
	if (
		parts.scheme == "https"
		and parts.netloc
		and not parts.username
		and not unquote(parts.path).replace("\\", "/").startswith("/private/")
	):
		return value
	return None


def project_product(product, *, certifications=(), price=None, commercial=False, configure_available=True):
	get = product.get
	family = FAMILY_ALIASES.get(get("product_type"), get("product_type"))
	template_field = TEMPLATE_FIELDS.get(family)
	template = get(template_field) if template_field else None
	active = parse_bool(get("is_active"))
	options, errors = [], []
	for row in get("configurator_options") or []:
		try:
			values = row.get("allowed_values_json") or []
			if isinstance(values, str):
				values = json.loads(values)
			if not isinstance(values, list) or any(not isinstance(v, (str, dict)) for v in values):
				raise ValueError("Allowed values must be an array of choices")
		except (ValueError, TypeError):
			values = []
			errors.append(
				{"field": "configurator_options", "code": "INVALID_OPTIONS", "step": row.get("option_step")}
			)
		options.append(
			{
				"step": row.get("option_step"),
				"type": row.get("option_type"),
				"label": row.get("option_label"),
				"description": row.get("option_description"),
				"required": parse_bool(row.get("is_required")),
				"depends_on_step": row.get("depends_on_step"),
				"allowed_values": values,
			}
		)
	capability = (
		"configure"
		if active and template and parse_bool(get("is_configurable")) and not errors and configure_available
		else "inquiry"
	)
	if not active:
		capability = "unavailable"
	configure_url = None
	if capability == "configure":
		configure_url = "/portal/configure?" + urlencode(
			{"category": family, "template": template, "product_slug": get("product_slug")}
		)
	documents = []
	for row in sorted(get("documents") or [], key=lambda r: r.get("display_order") or 0):
		url = safe_document_url(row.get("document_file"))
		if not url:
			continue
		documents.append(
			{
				"id": row.get("name"),
				"title": row.get("document_title"),
				"document_name": row.get("document_title"),
				"type": row.get("document_type"),
				"document_type": row.get("document_type"),
				"file_url": url,
				"download_url": url,
				"source": get("name"),
				"classification": "public_literature",
				"display_order": row.get("display_order") or 0,
			}
		)
	gallery = [
		{
			"image": row.get("image"),
			"alt_text": row.get("alt_text") or "",
			"display_order": row.get("display_order") or row.get("idx") or 0,
		}
		for row in get("gallery_images") or []
	]
	if not gallery and get("featured_image"):
		gallery = [
			{"image": get("featured_image"), "alt_text": get("product_name") or "", "display_order": 0}
		]
	result = {
		key: get(key)
		for key in (
			"name",
			"product_name",
			"product_slug",
			"product_type",
			"product_category",
			"series",
			"short_description",
			"featured_image",
			"dimensions_image",
			"features",
			"configurator_intro_text",
			"min_length_mm",
			"max_length_mm",
			"length_increment_mm",
			"fixture_template",
			"tape_neon_template",
			"led_sheet_template",
		)
	}
	result.update(
		{
			"projection_version": 1,
			"family": family,
			"is_active": active,
			"is_configurable": capability == "configure",
			"capability": capability,
			"configure_url": configure_url,
			"template": template,
			"gallery": sorted(gallery, key=lambda r: r["display_order"]),
			"documents": documents,
			"configurator_options": sorted(options, key=lambda r: r["step"] or 0),
			"validation_errors": errors,
			"certifications": list(certifications),
			"specifications": [
				{
					"spec_group": row.get("spec_group"),
					"spec_label": row.get("spec_label"),
					"spec_value": row.get("spec_value"),
					"spec_unit": row.get("spec_unit"),
					"display_order": row.get("display_order") or 0,
				}
				for row in sorted(get("specifications") or [], key=lambda r: r.get("display_order") or 0)
			],
			"feed_lengths": [
				{key: row.get(key) for key in ("step", "label", "code")} for row in get("feed_lengths") or []
			],
			"attribute_links": [
				{key: row.get(key) for key in ("attribute_type", "attribute_name", "display_label")}
				for row in get("attribute_links") or []
			],
			"compatible_products": [
				{key: row.get(key) for key in ("related_product", "relationship_type", "notes")}
				for row in get("compatible_products") or []
			],
		}
	)
	if commercial and family == "Linear Fixture" and price is not None:
		result["base_price_msrp"] = price
		result["pricing"] = {
			"basis": "fixture_base_msrp",
			"amount": price,
			"excludes": ["configuration-dependent components"],
		}
	return result

File: n8n_workflows/lib/product_projection.js
function projectWebflowProduct(product, brandConfig) {
// Transform ERPNext product to Webflow format - FULL DATA SYNC

const isUpdate = product.webflow_item_id ? true : false;

// ============================================
// FEATURE FLAGS
// ============================================
// Set to true AFTER creating the corresponding fields in your Webflow
// Products collection (CMS Designer).  Until then, keep false to avoid
// "Field not described in schema" 400 errors from Webflow.
const INCLUDE_ATTRIBUTE_DISPLAY_FIELDS = false; // e.g. finishes-5, cct-options-5
const INCLUDE_ATTRIBUTE_FILTER_FIELDS  = false; // e.g. finish-filter, cri-filter
const INCLUDE_CATEGORY_REF_FIELD       = false; // category-filter (ItemRef)

// ============================================
// URL HELPERS
// ============================================
const ERPNEXT_BASE_URL = brandConfig.erpnext_base_url;

const makeAbsoluteUrl = (url) => {
  if (!url) return null;
  let fullUrl = url;
  if (!url.startsWith('http://') && !url.startsWith('https://')) {
    fullUrl = url.startsWith('/') ? `${ERPNEXT_BASE_URL}${url}` : `${ERPNEXT_BASE_URL}/${url}`;
  }
  // Encode spaces and brackets in the path portion so Webflow can fetch the file
  const qIdx = fullUrl.indexOf('?');
  const path = qIdx >= 0 ? fullUrl.substring(0, qIdx) : fullUrl;
  const qs   = qIdx >= 0 ? fullUrl.substring(qIdx) : '';
  const encoded = path.replace(/ /g, '%20').replace(/\[/g, '%5B').replace(/\]/g, '%5D');
  return encoded + qs;
};


const isValidImageUrl = (url) => {
  if (!url) return false;
  if (url.includes('/private/')) return false;
  if (!url.startsWith('http://') && !url.startsWith('https://')) return false;
  return true;
};

const isValidFileUrl = (url) => {
  if (!url) return false;
  if (url.includes('/private/')) return false;
  return true;
};

// ============================================
// ATTRIBUTE FILTER FIELDS
// ============================================
// ERPNext provides pre-built plain-text filter data via filter_field_data.
// Filter fields contain comma-separated attribute names for Webflow CMS filtering.
const filterFieldData = product.filter_field_data || {};

// Plain text display fields (name, code formatted with pipe separators)
const attributeTextByType = product.attribute_text_by_type || {};
const attributeLinksByType = product.attribute_links_by_type || {};
const attributeLinks = product.attribute_links || [];

// Helper: get pipe-separated display text for an attribute type
const getAttrText = (type) => attributeTextByType[type] || '';

// Feed Direction - combine Feed Direction and Power Feed Type
const formatAttrPair = (a) => {
  const label = a.display_label || a.attribute_name || '';
  const code = a.attribute_code || '';
  if (!label) return '';
  return code ? `${label}, ${code}` : label;
};
const feedDirectionPairs = (attributeLinksByType['Feed Direction'] || []).map(formatAttrPair).filter(v => v);
const powerFeedTypePairs = (attributeLinksByType['Power Feed Type'] || []).map(formatAttrPair).filter(v => v);
const allFeedDirectionText = [...new Set([...feedDirectionPairs, ...powerFeedTypePairs])].join(' | ');

// LED Package
const ledPackageAttrs = attributeLinksByType['LED Package'] || [];
const fixtureTypes = ledPackageAttrs.map(formatAttrPair).filter(v => v).join(' | ');

// ============================================
// GALLERY IMAGES (Multi-image field)
// ============================================
const rawGalleryImages = product.gallery_images || [];
const galleryImages = rawGalleryImages
  .map(img => {
    const imageUrl = makeAbsoluteUrl(img.image || img.url || img.file_url);
    return {
      url: imageUrl,
      alt: img.alt_text || img.caption || product.product_name
    };
  })
  .filter(img => isValidImageUrl(img.url));

// ============================================
// DOCUMENTS (JSON for file links)
// ============================================
const documents = (product.documents || []).map(doc => ({
  url: makeAbsoluteUrl(doc.document_file),
  type: doc.document_type,
  title: doc.document_title,
  order: doc.display_order
}));

// ============================================
// SPEC SHEET FILE (for Webflow file field)
// ============================================
const specSheetDoc = (product.documents || []).find(doc => doc.document_type === 'Spec Sheet');
const specSheetUrl = specSheetDoc ? makeAbsoluteUrl(specSheetDoc.document_file) : null;

// ============================================
// CERTIFICATIONS
// ============================================
const certifications = (product.certifications || []).map(c => ({
  name: c.certification,
  code: c.certification_details?.certification_code,
  body: c.certification_details?.certification_body,
  badge: makeAbsoluteUrl(c.certification_details?.badge_image)
}));

// ============================================
// COMPATIBLE PRODUCTS
// ============================================
const compatibleProducts = (product.compatible_products || []).map(cp => ({
  product: cp.related_product,
  type: cp.relationship_type,
  notes: cp.notes
}));

// ============================================
// CONFIGURATOR OPTIONS
// ============================================
const configuratorOptions = (product.configurator_options || []).map(opt => ({
  step: opt.option_step,
  type: opt.option_type,
  label: opt.option_label,
  description: opt.option_description,
  required: opt.is_required,
  dependsOn: opt.depends_on_step,
  allowedValues: opt.allowed_values_json
}));

// ============================================
// FEED LENGTHS (Webflow part-number configurator)
// ============================================
const feedLengths = (product.feed_lengths || []).map(fl => ({
  step: fl.step,
  label: fl.label,
  code: fl.code
}));

// ============================================
// KIT COMPONENTS
// ============================================
const kitComponents = (product.kit_components || []).map(k => ({
  type: k.component_type,
  item: k.component_item,
  specDoctype: k.component_spec_doctype,
  specName: k.component_spec_name,
  qty: k.quantity,
  notes: k.notes
}));

// ============================================
// LENGTH CONVERSIONS
// ============================================
const mmToInches = (mm) => mm ? Math.round((mm / 25.4) * 100) / 100 : 0;

// ============================================
// BUILD FIELD DATA
// ============================================
const fieldData = {
  // Basic Info
  "name": product.product_name,
  "product-type": product.product_type,
  "short-description": product.short_description || '',
  "subheading": product.sublabel || '',
  "erp-sync-id": product.product_slug,
  
  // Configurator Settings - NOTE: field is "is-configurable-3" in Webflow
  "is-configurable-3": product.is_configurable === 1 || product.is_configurable === true,
  "configurator-intro": product.configurator_intro_text || '',
  "min-length-mm": product.min_length_mm || 0,
  "max-length-mm": product.max_length_mm || 0,
  "min-length-inches": mmToInches(product.min_length_mm),
  "max-length-inches": mmToInches(product.max_length_mm),
  
  // Series Info (from linked series doctype)
  "series-display-name": product.series || product.series_display_name || '',
  
  // Sync timestamp

  
  // JSON data fields
  "specifications-json": JSON.stringify(product.specifications || []),
  "attribute-links-json": JSON.stringify(attributeLinks),
  "documents-json": JSON.stringify(documents),
  "certifications-json": JSON.stringify(certifications),
  "configurator-options-json": JSON.stringify(configuratorOptions),
  "feed-lengths-json": JSON.stringify(feedLengths),
  "kit-components-json": JSON.stringify(kitComponents),
  "compatible-products-json": JSON.stringify(compatibleProducts),
  // Features JSON
  "features-json": product.features_json || "[]",
  // Product Badge
  "product-badge": product.product_badge || "",
};

// ============================================
// CATEGORY REFERENCE (requires Webflow Item ID)
// ============================================
const categoryWebflowId = product.category_webflow_item_id || product.product_category_webflow_item_id || (product.category_details && product.category_details.webflow_item_id);
if (INCLUDE_CATEGORY_REF_FIELD && categoryWebflowId && /^[a-f0-9]{24}$/i.test(categoryWebflowId)) {
  fieldData["category-filter"] = categoryWebflowId;
}

// ============================================
// PLAIN TEXT ATTRIBUTE FIELDS (existing, unchanged)
// ============================================
// These plain text fields remain for display purposes.
if (INCLUDE_ATTRIBUTE_DISPLAY_FIELDS) {
fieldData['finishes-5'] = getAttrText('Finish');
fieldData['lens-options-5'] = getAttrText('Lens Appearance');
fieldData['mounting-methods-5'] = getAttrText('Mounting Method');
fieldData['cct-options-5'] = getAttrText('CCT');
fieldData['output-levels-5'] = getAttrText('Output Level');
fieldData['cris-5'] = getAttrText('CRI');
fieldData['feed-directions-5'] = allFeedDirectionText;
fieldData['environment-ratings-5'] = getAttrText('Environment Rating');
fieldData['fixture-types-5'] = fixtureTypes;
}

// ============================================
// PLAIN TEXT FILTER FIELDS (for Webflow CMS filtering)
// ============================================
// These fields (with '-filter' suffix) contain comma-separated attribute
// names for Webflow CMS filtering. They are plain text, not multi-reference.
if (INCLUDE_ATTRIBUTE_FILTER_FIELDS) {
for (const [slug, text] of Object.entries(filterFieldData)) {
  if (text) {
    fieldData[slug] = text;
  }
}
}

// Slug only on create
if (!isUpdate) {
  fieldData["slug"] = product.product_slug;
}

// ============================================
// BUILD WEBFLOW PAYLOAD
// ============================================
const webflowData = {
  isArchived: false,
  isDraft: true,
  fieldData: fieldData
};

// ============================================
// IMAGE FIELDS
// ============================================
const featuredImageUrl = makeAbsoluteUrl(product.featured_image);
if (isValidImageUrl(featuredImageUrl)) {
  webflowData.fieldData["featured-image"] = { url: featuredImageUrl };
}

const dimensionsImageUrl = makeAbsoluteUrl(product.dimensions_image);
if (isValidImageUrl(dimensionsImageUrl)) {
  webflowData.fieldData["dimensions"] = { url: dimensionsImageUrl };
}

const seriesFamilyImageUrl = makeAbsoluteUrl(product.series_family_image);
if (isValidImageUrl(seriesFamilyImageUrl)) {
  webflowData.fieldData["family-image"] = { url: seriesFamilyImageUrl };
}

if (galleryImages.length > 0) {
  webflowData.fieldData["gallery"] = galleryImages;
}

// ============================================
// SPEC SHEET FILE FIELD
// ============================================
if (specSheetUrl && isValidFileUrl(specSheetUrl)) {
  webflowData.fieldData["spec-sheet"] = { url: specSheetUrl };
}

// ============================================
// PER-BRAND CONFIGURATOR GUARD (defense in depth — server also strips)
// ============================================
try {

  if (brandConfig && brandConfig.include_configurator_payload === false) {
    delete webflowData.fieldData['is-configurable-3'];
    delete webflowData.fieldData['configurator-intro'];
    delete webflowData.fieldData['min-length-mm'];
    delete webflowData.fieldData['max-length-mm'];
    delete webflowData.fieldData['min-length-inches'];
    delete webflowData.fieldData['max-length-inches'];
    delete webflowData.fieldData['configurator-options-json'];
    delete webflowData.fieldData['feed-lengths-json'];
    delete webflowData.fieldData['kit-components-json'];
  }
} catch (e) { throw e; }

// ============================================
// RETURN FULL DATA
// ============================================
return {
  json: {
    webflowData,
    originalProduct: product,
    _debug: {
      featuredImageUrl,
      dimensionsImageUrl,
      galleryImagesCount: galleryImages.length,
      rawGalleryCount: rawGalleryImages.length,
      rawGalleryImages: rawGalleryImages.slice(0, 3),
      processedGalleryImages: galleryImages.slice(0, 3),
      fixtureTypes: fixtureTypes,
      ledPackageAttrs: ledPackageAttrs,
      categoryWebflowId: categoryWebflowId,
      productCategory: product.product_category,
      categoryDetails: product.category_details,
      specSheetUrl: specSheetUrl,
      filterFieldData: filterFieldData,
      filterSlugsUsed: Object.keys(filterFieldData).filter(s => filterFieldData[s])
    }
  }
};
}
if (typeof module !== "undefined") module.exports = { projectWebflowProduct };
